
''' coverage rules '''
#def scale_factor(wildcards):
#    ''' parse the alignment log to get the number aligned reads '''
#    align_log = 'logs/align.%s.log.err' % wildcards['sample']
#    with open(align_log) as logfile:
#        print 'blah'

rule end_coverage:
    input:
        'alignment/{sample}/{sample}.bam'
    output:
        'coverage/{sample}/{sample}.{side}.{strand}.bg.gz'
    params:
        chromsize = '/vol3/home/jhessel/ref/genomes/sacCer1/sacCer1.chrom.sizes',
        side_arg = lambda wildcards: SIDES[wildcards.side],
        strand_arg = lambda wildcards: STRANDS[wildcards.strand],
        job_name = '{sample}.{side}.{strand}.coverage',
#         scale = scale_factor,
        doc = "determine coverage of 5p and 3p ends, for each strand"
    log: 
        'logs/coverage.{sample}.log'
    shell:
        "bedtools genomecov -ibam {input} -g {params.chromsize} -bg "
        "{params.strand_arg} {params.side_arg} "
        "| gzip -c > {output}"


def merge_coverage_input(wildcards):
    filebase = 'coverage/{sample}/{sample}.{side}.{strand}.bg.gz'
    return [filebase.format(sample = wildcards.sample,
                            strand = wildcards.strand,
                            side = side) for side in SIDES]


rule merge_coverage:
    input: merge_coverage_input
    output:
        'coverage/{sample}/{sample}.combined.{strand}.bg.gz'
    params:
        job_name = '{sample}.{strand}.merge_coverage',
        doc = "combine coverage of ends into a merged bedgraph file"
    shell:
        "bedtools unionbedg -i {input} "
        "| awk '{{print $1, $2, $3, $4 + $5}}' "
        "| gzip -c > {output}"



